### ISSUE TITLE
Generic on Ix dim

### ISSUE THREAD

**Issue by nilgoyette** [rust-ndarray/ndarray#1243]
I'm trying to code a function that returns the centered indices of a n-d kernel (structuring element) when the value is true. I hardcoded it for 3-D because I don't know how to make it generic

    fn build_indices(kernel: ArrayView3<bool>, radii: &[usize]) -> Vec<[isize; 3]> {
        let radii = [radii[0] as isize, radii[1] as isize, radii[2] as isize];
        kernel
            .indexed_iter()
            .filter_map(|(idx, &b)| {
                b.then_some([
                    idx.0 as isize - radii[0],
                    idx.1 as isize - radii[1],
                    idx.2 as isize - radii[2],
                ])
            })
            .collect()
    }

As you can guess, I want to return `Vec<[isize; NB_DIM]>` instead. But an index is a tuple in ndarray and I don't know how to work or transform them. I tried adding `where D: Dimension` and calling `kernel.raw_dim()` to get the `D`, which should be `Ix2` or `Ix3`, etc. But then I'm unable to do anything useful with that object. I can't access `ix()` or the actual constant, like 2, behind `Ix2`.

What would you suggest?

**Comment by adamreichold** [comment]
Does 

```rust
fn build_indices<const N: usize>(
    kernel: ArrayView<bool, Dim<[usize; N]>>,
    radii: [usize; N],
) -> Vec<[isize; N]>
where
    Dim<[usize; N]>: Dimension,
{
    let radii = radii.map(|radius| radius as isize);
    kernel
        .indexed_iter()
        .filter_map(|(idx, &b)| {
            b.then_some({
                let idx = idx.into_dimension();
                let mut res = [0; N];
                for n in 0..N {
                    res[n] = idx[n] as isize - radii[n];
                }
                res
            })
        })
        .collect()
}
```

work in your intended use case?

I think the main issue here is that `indexed_iter` explicitly returns the pattern-matching friendly tuple form of `Dim` (This was probably before one could pattern match on arrays.) and hence one has to explicit turn into back into the indexable variant using the `IntoDimension` trait. Hopefully, the compiler will optimize away this back and forth.

**Comment by nilgoyette** [comment]
where Dim<[usize; N]>: Dimension,

I wasn't aware of this syntax. I always saw the `ArrayBase<S, D>` `D: Dimension` before.

I didn't test it yet, but I'm pretty sure this solves my problem. Thank you!

**Comment by adamreichold** [comment]
> I wasn't aware of this syntax. I always saw the ArrayBase<S, D> D: Dimension before.

Yeah, this is required because we currently do not implement `Dimension` for all `N` (due to lack of const generics) and it basically constrains `N` to those values for which we do. And indeed, the left hand side of a trait is not restricted to just a type variable.

**Comment by nilgoyette** [comment]
@adamreichold If I may, I have a followup question.

In the context of a "normal" generic function `ArrayBase<S, D>` `D: Dimension`, can I call a function that uses `const N: usize`? I get this error

    arguments to this function are incorrect
    expected reference `&ArrayBase<_, Dim<[usize; _]>>`
       found reference `&ArrayBase<SM, D>`
    expected struct `ArrayBase<ViewRepr<&bool>, Dim<[usize; _]>>`
       found struct `ArrayBase<ViewRepr<&bool>, D>`

I agree with the message... the types are indeed different. But I would have thought that two functions "generic over the dimension" could call each other.

Am I forced to use `const N: usize` everywhere now that I used it at one place?
